### PR TITLE
spec updates

### PR DESCRIPTION
**Pull request by arnetheduck**
* lots of renames
* add some epoch processing

### CODE REVIEW

File: beacon_chain/state_transition.nim [hunk 1]
```diff
@@ -127,20 +158,199 @@ func process_block*(state: BeaconState, blck: BeaconBlock): Option[BeaconState]
...
+func adjust_for_inclusion_distance[T](magnitude: T, dist: T): T =
+  magnitude div 2 + (magnitude div 2) * MIN_ATTESTATION_INCLUSION_DELAY div dist
+
+func processEpoch*(state: BeaconState, blck: BeaconBlock): Option[BeaconState] =
+  ## Epoch processing happens every time we've passed EPOCH_LENGTH blocks.
+  ## Because some slots may be skipped, it may happen that we go through the
+  ## loop more than once - each time the latest_state_recalculation_slot will be
+  ## increased by EPOCH_LENGTH.
+
+  # TODO: simplistic way to be able to rollback state
```
Comment (mratsim) [inline beacon_chain/state_transition.nim]: as in Eth1.0, we have a candidate state that we build during processing and commit it all at once. This reduce the chance of an inconsistent state, for example due to a forced shutdown.

File: beacon_chain/state_transition.nim [hunk 2]
```diff
@@ -127,20 +158,199 @@ func process_block*(state: BeaconState, blck: BeaconBlock): Option[BeaconState]
...
+        mapIt(previous_epoch_boundary_attesters,
+          get_effective_balance(state.validator_registry[it]))
+      )
+
+    # TODO gets pretty hairy here
+    func attesting_validators(
+        obj: ShardAndCommittee, shard_block_hash: Eth2Digest): seq[int] =
+      flatten(
+        mapIt(
+          filterIt(concat(this_epoch_attestations, previous_epoch_attestations),
```
Comment (mratsim) [inline beacon_chain/state_transition.nim]: Do we have to use closures here?

File: beacon_chain/state_transition.nim [hunk 3]
```diff
@@ -127,20 +158,199 @@ func process_block*(state: BeaconState, blck: BeaconBlock): Option[BeaconState]
...
+    # Let `inclusion_slot(v)` equal `a.slot_included` for the attestation `a` where `v` is in `get_attestation_participants(state, a.data, a.participation_bitfield)`, and `inclusion_distance(v) = a.slot_included - a.data.slot` for the same attestation. We define a function `adjust_for_inclusion_distance(magnitude, distance)` which adjusts the reward of an attestation based on how long it took to get included (the longer, the lower the reward). Returns a value between 0 and `magnitude`.
+
+    # Adjust justified slots and crosslink status
+
+    var new_justified_slot: Option[uint64]
+    # overflow intentional!
+    state.justified_slot_bitfield = state.justified_slot_bitfield * 2
+
+    if 3'u64 * previous_epoch_boundary_attesting_balance >= 2'u64 * total_balance:
+      # TODO spec says "flip the second lowest bit to 1" and does "AND", wrong?
```
Comment (mratsim) [inline beacon_chain/state_transition.nim]: Proper bitfield needed in the specs with its serialization format and we won't have those issues.

File: beacon_chain/state_transition.nim (hunk 1 above)
Comment (arnetheduck) [reply beacon_chain/state_transition.nim]: yeah, I'm not married to this way of doing it, just wanna keep it in here as a reminder - in the future I suspect we're going to have to keep multiple states around for potential forks etc

File: beacon_chain/state_transition.nim (hunk 2 above)
Comment (arnetheduck) [reply beacon_chain/state_transition.nim]: for now, yeah, in the spec it precomputes a bunch of variables (`this_epoch_attestations` and friends), and then uses them for lookups - it's either closures or helper functions where we explicitly pass them.. I suspect the code can be simplified quite a bit once it settles down.

File: beacon_chain/state_transition.nim (hunk 3 above)
Comment (arnetheduck) [reply beacon_chain/state_transition.nim]: yeah, this bit here is a finalized-block-detector that's a bit weird

**Pull request merged**